client/src/main/java/io/split/engine/experiments/SplitFetcherImp.java

```diff
@@ -95,36 +116,55 @@ public void run() {
     }
 
     private Set<String> runWithoutExceptionHandling(FetchOptions options) throws InterruptedException, UriTooLongException {
-        SplitChange change = _splitChangeFetcher.fetch(_splitCacheProducer.getChangeNumber(), options);
+        SplitChange change = _splitChangeFetcher.fetch(_splitCacheProducer.getChangeNumber(),
+                _ruleBasedSegmentCacheProducer.getChangeNumber(), options);
         Set<String> segments = new HashSet<>();
 
         if (change == null) {
             throw new IllegalStateException("SplitChange was null");
         }
 
-        if (change.since != _splitCacheProducer.getChangeNumber() || change.till < _splitCacheProducer.getChangeNumber()) {
-            // some other thread may have updated the shared state. exit
+        if (checkExitConditions(change.featureFlags, _splitCacheProducer.getChangeNumber()) ||
+            checkExitConditions(change.ruleBasedSegments, _ruleBasedSegmentCacheProducer.getChangeNumber())) {
```

Comment:
i am wondering if the exit conditions should be use`&&` and not `||`, because if someone is false, that means that is still pending some changes.

Reply:
```suggestion
        if (checkExitConditions(change.featureFlags, _splitCacheProducer.getChangeNumber()) &&
            checkExitConditions(change.ruleBasedSegments, _ruleBasedSegmentCacheProducer.getChangeNumber())) {
```

Reply:
The condition should be: or, the condition is checking if we need to continue fetching based on the received change number.
If we make it (and), then if splits needs more fetching, but  rbs does not, it will exit, and we never fetch the latest split changes.
If we make it (or), then the fetching will continue if either of rbs or splits need fetching.